File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Bucket sort
        counts = {}
        freq = [[] for i in range(len(nums)+1)]
        res = []

        for num in nums:
            counts[num] = 1 + counts.get(num, 0)
        for num, count in counts.items():
            freq[count].append(num)

        for i in range(len(freq)-1, 0, -1):
            for num in freq[i]:
                res.append(num)
                if len(res) == k:
                    return res

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // # Bucket sort
        unordered_map<int, int> counts;
        vector<vector<int>> freq(nums.size() + 1);
        vector<int> res;

        for (auto& num: nums)
        {
            counts[num] += 1;
        }
        for (auto& key_val: counts)
        {
            freq[key_val.second].push_back(key_val.first);
        }

        for (int i = freq.size() - 1 ; i >= 0 ; i--)
        {
            for (auto& num: freq[i])
            {
                res.push_back(num);
                if (res.size() == k)
                {
                    return res;
                }
            }
        }
        return {};
    }
};